diff --git a/kernel/policy/allowlist.c b/kernel/policy/allowlist.c
index d5eac53f..00000000 100644
--- a/kernel/policy/allowlist.c
+++ b/kernel/policy/allowlist.c
@@ -3,6 +3,7 @@
 #include <linux/rculist.h>
 #include <linux/mutex.h>
 #include <linux/task_work.h>
+#include <linux/sched/task.h>
 #include <linux/capability.h>
 #include <linux/compiler.h>
 #include <linux/fs.h>
@@ -13,6 +14,12 @@
 #include <linux/slab.h>
 #include <linux/types.h>
 #include <linux/version.h>
+
+#if LINUX_VERSION_CODE < KERNEL_VERSION(5, 0, 0)
+#define KSU_TWA_RESUME true
+#else
+#define KSU_TWA_RESUME TWA_RESUME
+#endif
 #include <linux/compiler_types.h>
 #include <linux/hashtable.h>
 #include <linux/kref.h>
@@ -485,7 +492,7 @@ void ksu_persistent_allow_list()
         goto put_task;
     }
     cb->func = do_persistent_allow_list;
-    if (task_work_add(tsk, cb, TWA_RESUME)) {
+    if (task_work_add(tsk, cb, KSU_TWA_RESUME)) {
         kfree(cb);
         pr_warn("save_allow_list add task_work failed\n");
     }
