diff --git a/kernel/infra/seccomp_cache.c b/kernel/infra/seccomp_cache.c
index a9d2b634..00000000 100644
--- a/kernel/infra/seccomp_cache.c
+++ b/kernel/infra/seccomp_cache.c
@@ -5,6 +5,12 @@
 #include <linux/uaccess.h>
 #include <linux/filter.h>
 #include <linux/seccomp.h>
+
+#ifndef SECCOMP_ARCH_NATIVE_NR
+#include <asm/unistd.h>
+#define SECCOMP_ARCH_NATIVE_NR __NR_syscalls
+#endif
+
 #include "klog.h" // IWYU pragma: keep
 #include "infra/seccomp_cache.h"
 
