diff --git a/kernel/policy/app_profile.c b/kernel/policy/app_profile.c
--- a/kernel/policy/app_profile.c
+++ b/kernel/policy/app_profile.c
@@ -63,7 +63,9 @@ void setup_groups(struct root_profile *profile, struct cred *cred)
     put_group_info(group_info);
 }

+#if LINUX_VERSION_CODE >= KERNEL_VERSION(5, 9, 0)
 void seccomp_filter_release(struct task_struct *tsk);
+#endif

 static void disable_seccomp(void)
 {
@@ -100,7 +102,9 @@ static void disable_seccomp(void)
     fake->sighand = NULL;
 #endif

+#if LINUX_VERSION_CODE >= KERNEL_VERSION(5, 9, 0)
     seccomp_filter_release(fake);
+#endif
     kfree(fake);
 }
